interp/src/utils.rs

```diff
@@ -47,6 +47,36 @@ impl From<&RRC<Port>> for PortRef {
     }
 }
 
+pub(super) struct PortAssignment<'a>(*const Port, &'a Assignment);
+
+impl<'a> Hash for PortAssignment<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl<'a> PartialEq for PortAssignment<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.0, other.0)
+    }
+}
+
+impl<'a> Eq for PortAssignment<'a> {}
+
+impl<'a, 'b> PortAssignment<'a> {
+    pub fn new(p_ref: &'b Port, a_ref: &'a Assignment) -> Self {
+        Self(p_ref as *const Port, a_ref)
+    }
+
+    pub fn get_port(&self) -> &Port {
+        unsafe { self.0.as_ref().unwrap() }
```

Comment:
Why does this need to be `unsafe`?

Reply:
I think it is because the first field of this struct is of raw-pointer type `*const ir::Port`?

Reply:
Indeed; dereferencing "raw pointers" is unsafe (because we don't know that the referent still exists). I suppose the broader question is whether there is a way around it… that is, could `PortAssignment` just contain a `&'a Port` instead of a `*const Port`? That would avoid the need for raw pointer juggling here altogether.